README
NAME
    WWW::Proxy4FreeCom - fetch proxy list from http://proxy4free.com/

SYNOPSIS
        use strict;
        use warnings;
        use WWW::Proxy4FreeCom;

        my $prox = WWW::Proxy4FreeCom->new;

        my $proxies = $prox->get_list
            or die $prox->error;

        printf "%-40s (last tested %s ago)\n", @$_{ qw(domain last_test) }
            for @$proxies;

DESCRIPTION
    The module provides means to fetch proxy list from
    <http://proxy4free.com/> website.

CONSTRUCTOR
  "new"
        my $prox = WWW::Proxy4FreeCom->new;

        my $prox = WWW::Proxy4FreeCom->new(
            timeout => 10,
            debug   => 1,
        );

        my $prox = WWW::Proxy4FreeCom->new(
            ua => LWP::UserAgent->new(
                timeout => 10,
                agent   => 'ProxUA',
            ),
        );

    Constructs and returns a brand new yummy juicy WWW::Proxy4FreeCom
    object. Takes a few *optional* arguments. Possible arguments are as
    follows:

   "timeout"
        ->new( timeout => 10 );

    Optional. Specifies the "timeout" argument of LWP::UserAgent's
    constructor, which is used for retrieving data. Defaults to: 30 seconds.

   "ua"
        ->new( ua => LWP::UserAgent->new(agent => 'Foos!') );

    Optional. If the "timeout" argument is not enough for your needs of
    mutilating the LWP::UserAgent object used for retrieving proxy list,
    feel free to specify the "ua" argument which takes an LWP::UserAgent
    object as a value. Note: the "timeout" argument to the constructor will
    not do anything if you specify the "ua" argument as well. Defaults to:
    plain boring default LWP::UserAgent object with "timeout" argument set
    to whatever "WWW::Proxy4FreeCom"'s "timeout" argument is set to as well
    as "agent" argument is set to mimic Firefox.

   "debug"
        ->new( debug => 1 );

    When "get_list()" is called any unsuccessfull page retrievals will be
    silently ignored. Setting "debug" argument to a true value will "carp()"
    any network errors if they occur.

METHODS
  "get_list"
        my $list_ref = $prox->get_list # just from the "proxy list 1"
            or die $prox->error;

        my $list_ref = $prox->get_list( 2 ) # just from the "proxy list 2"
            or die $prox->error;

        $prox->get_list( [3,5] ) # lists 3 and 5 only
            or die $prox->error;

    Instructs the objects to fetch a fresh list of proxies from
    <http://proxy4free.com/>. On failure returns "undef" or an empty list,
    depending on the context, and the human-readable error will be available
    by calling the "->error" method. On success returns an arrayref of
    hashrefs, each representing a proxy entry. Takes one optional argument
    which can be either a number between 1 and 14 (inclusive) or an arrayref
    with several of these numbers. The numbers represent the page number of
    proxy list pages on <http://proxy4free.com/>. By default only the list
    from the "proxy list 1" will be fetched.

    Each hashref in the returned arrayref is in a following format (if any
    field is missing on the site it will be reported as a string "N/A"):

        {
            'domain' => 'localfast.info',
            'rating' => '65',
            'country' => 'Germany',
            'access_time' => '1.3',
            'uptime' => '96',
            'online_since' => '16 hours',
            'last_test' => '30 minutes',
            'features_hian' => '1',
            'features_ssl' => '0',
        }

    Where all the values correspond to the proxy list table columns on the
    website. The "features_hian" and "features_ssl" keys will be set to true
    values, if the proxy offers "HiAn" or "SSL" features respectively.

  "error"
        my $list = $prox->get_list # just from the "proxy list 1"
            or die $prox->error;

    If "get_list()" method fails it will return either "undef" or an empty
    list, depending on the context, and the reason for the error will be
    available via "error()" method. Takes no arguments, return a
    human-readable error message explaining the failure.

  "list"
        my $last_list_ref = $prox->list;

    Must be called after a successfull call to "get_list()". Takes no
    arguments, returns the same arrayref of hashrefs as last call to
    "get_list()" returned.

  "ua"
        my $old_LWP_UA_obj = $prox->ua;

        $prox->ua( LWP::UserAgent->new( timeout => 10, agent => 'foos' );

    Returns a currently used LWP::UserAgent object used for retrieving data.
    Takes one optional argument which must be an LWP::UserAgent object, and
    the object you specify will be used in any subsequent calls to
    "get_list()".

  "debug"
        my $old_debug => $prox->debug;

        $prox->debug(1);

    Returns a currently set debug value, when called with an optional
    argument (which can be either a true or false value) will set debug to
    that value. See "debug" argument to constructor for more information.

AUTHOR
    'Zoffix, "<'zoffix at cpan.org'>" (<http://zoffix.com>,
    <http://haslayout.net>)

BUGS
    Please report any bugs or feature requests to "bug-www-proxy4freecom at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Proxy4FreeCom>. I
    will be notified, and then you'll automatically be notified of progress
    on your bug as I make changes.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc WWW::Proxy4FreeCom

    You can also look for information at:

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Proxy4FreeCom>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/WWW-Proxy4FreeCom>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/WWW-Proxy4FreeCom>

    *   Search CPAN

        <http://search.cpan.org/dist/WWW-Proxy4FreeCom>

COPYRIGHT & LICENSE
    Copyright 2008 'Zoffix, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.



Changes
Revision history for WWW-Proxy4FreeCom

1.003   2014-01-04
        - Updated tests to handle an edge case for country key.

1.002   2013-12-30
        - Fixed broken tests

1.001   2013-12-27
        [ Major BREAKING Changes ]
        - This version has a completely new API, due to the website update.
          Previous version was completely broken.
        - Filtering (->filter) was removed altogether.

        [ Minor Changes]
        - Minor fixes to comply with the specs and Kwalitee game
        - Added autogeneration of README from POD
        - Moved author tests to xt/
        - Switched away from now deprecated Class::Data::Accessor
          to Class::Accessor::Grouped

0.001   2008-03-14
        - First version, released on an unsuspecting world.



Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    module_name         => 'WWW::Proxy4FreeCom',
    license             => 'perl',
    dist_author         => 'Zoffix <zoffix@cpan.org>',
    dist_version_from   => 'lib/WWW/Proxy4FreeCom.pm',
    build_requires => {
        'Test::More' => 0,
    },
    requires => {
        'Carp'                     => 1.04,
        'URI'                      => 1.35,
        'LWP::UserAgent'           => 2.036,
        'Mojo::DOM'                => 0,
        'Class::Accessor::Grouped' => 0,
        'perl'                     => '5.006',
    },
    configure_requires => { 'Module::Build' => 0 },
    add_to_cleanup      => [ 'WWW-Proxy4FreeCom-*' ],
    create_makefile_pl => 'traditional',
);
if ( $ENV{RELEASE_TESTING} ) {
    system('pod2text lib/WWW/Proxy4FreeCom.pm > README');
}

$builder->create_build_script();


lib/WWW/Proxy4FreeCom.pm
package WWW::Proxy4FreeCom;

use warnings;
use strict;

our $VERSION = '1.003';

use Carp;
use URI;
use LWP::UserAgent;
use Mojo::DOM;

use base 'Class::Accessor::Grouped';
__PACKAGE__->mk_group_accessors( simple => qw/
    list
    error
    ua
    debug
/);

sub new {
    my $self = bless {}, shift;

    croak "Must have even number of arguments to new()"
        if @_ & 1;
    my %args = @_;
    $args{ +lc } = delete $args{ $_ } for keys %args;

    $args{timeout} ||= 30;

    $args{ua} ||= LWP::UserAgent->new(
            timeout => $args{timeout},
            agent   => 'Mozilla/5.0 (X11; Ubuntu; Linux i686; rv:26.0)'
                        .' Gecko/20100101 Firefox/26.0',
    );

    $self->ua( $args{ua} );
    $self->debug( $args{debug} );

    return $self;
}

sub get_list {
    my $self = shift;
    my $custom_pages = shift;

    $self->$_(undef) for qw(list error);

    my @pages_list
        = defined $custom_pages
        ?  ( ref $custom_pages ? @$custom_pages : $custom_pages )
        : ( 1 );

    return $self->_set_error('Page number can only be 1..14')
        if grep { $_ < 1 or $_ > 14 } @pages_list;

    my $ua = $self->ua;
    my @proxies;
    for ( @pages_list ) {
        my $response = $ua->get(
            'http://www.proxy4free.com/list/webproxy' . $_ . '.html'
        );

        if ( $response->is_success ) {
            my $parse = $self->_parse_proxy_list(
                $response->decoded_content
            ) or return; ## parse error; error is already set

            push @proxies, @$parse;
        }
        else {
            $self->debug
                and carp "Page $_: " . $response->status_line;
        }
    }

    return $self->list( \@proxies );
}

sub _parse_proxy_list {
    my ( $self, $content ) = @_;

    my $dom = Mojo::DOM->new( $content );
    my @proxies;
    eval {
        for my $tr ( $dom->find('.proxy-list tbody tr')->each ) {
            my %tds;
            @tds{qw/domain  features_hian  features_ssl/}
            = ( $tr->find('td')->each )[1, 9, 10];

            @tds{qw/
                rating  country  access_time
                uptime  online_since  last_test
            /} = map "$_", map $_->text, ( $tr->find('td')->each )[ 3..8 ];

            $tds{domain} = $tds{domain}->find('a')->text;
            $tds{ $_ } = $tds{ $_ } =~ /on/ ? 1 : 0
                for qw/features_hian  features_ssl/;

            $_ = "$_" for values %tds;

            push @proxies, +{ %tds };
        }
    };

    $@ and return $self->_set_error("Parser error: $@");

    return \@proxies;
}

sub _set_error {
    my ( $self, $error_or_response, $type ) = @_;
    if ( defined $type and $type eq 'net' ) {
        $self->error( 'Network error: ' . $error_or_response->status_line );
    }
    else {
        $self->error( $error_or_response );
    }
    return;
}

1;
__END__

=encoding utf8

=head1 NAME

WWW::Proxy4FreeCom - fetch proxy list from http://proxy4free.com/

=head1 SYNOPSIS

    use strict;
    use warnings;
    use WWW::Proxy4FreeCom;

    my $prox = WWW::Proxy4FreeCom->new;

    my $proxies = $prox->get_list
        or die $prox->error;

    printf "%-40s (last tested %s ago)\n", @$_{ qw(domain last_test) }
        for @$proxies;

=head1 DESCRIPTION

The module provides means to fetch proxy list
from L<http://proxy4free.com/> website.

=head1 CONSTRUCTOR

=head2 C<new>

    my $prox = WWW::Proxy4FreeCom->new;

    my $prox = WWW::Proxy4FreeCom->new(
        timeout => 10,
        debug   => 1,
    );

    my $prox = WWW::Proxy4FreeCom->new(
        ua => LWP::UserAgent->new(
            timeout => 10,
            agent   => 'ProxUA',
        ),
    );

Constructs and returns a brand new yummy juicy WWW::Proxy4FreeCom
object. Takes a few I<optional> arguments. Possible arguments are
as follows:

=head3 C<timeout>

    ->new( timeout => 10 );

B<Optional>. Specifies the C<timeout> argument of L<LWP::UserAgent>'s
constructor, which is used for retrieving data.
B<Defaults to:> C<30> seconds.

=head3 C<ua>

    ->new( ua => LWP::UserAgent->new(agent => 'Foos!') );

B<Optional>. If the C<timeout> argument is not enough for your needs
of mutilating the L<LWP::UserAgent> object used for retrieving proxy list,
feel free
to specify the C<ua> argument which takes an L<LWP::UserAgent> object
as a value. B<Note:> the C<timeout> argument to the constructor will
not do anything if you specify the C<ua> argument as well. B<Defaults to:>
plain boring default L<LWP::UserAgent> object with C<timeout> argument
set to whatever C<WWW::Proxy4FreeCom>'s C<timeout> argument is
set to as well as C<agent> argument is set to mimic Firefox.

=head3 C<debug>

    ->new( debug => 1 );

When C<get_list()> is called any unsuccessfull page retrievals will be
silently ignored. Setting C<debug> argument to a true value will C<carp()>
any network errors if they occur.

=head1 METHODS

=head2 C<get_list>

    my $list_ref = $prox->get_list # just from the "proxy list 1"
        or die $prox->error;

    my $list_ref = $prox->get_list( 2 ) # just from the "proxy list 2"
        or die $prox->error;

    $prox->get_list( [3,5] ) # lists 3 and 5 only
        or die $prox->error;

Instructs the objects to fetch a fresh list of proxies from
L<http://proxy4free.com/>. B<On failure> returns C<undef> or an
empty list, depending on the context, and the human-readable error
will be available by calling the C<< ->error >> method.
B<On success> returns an arrayref of
hashrefs, each representing a proxy entry. Takes one optional argument which
can be either a number between 1 and 14 (inclusive) or an arrayref with
several of these numbers. The numbers represent the page number of
proxy list pages on L<http://proxy4free.com/>.
B<By default> only the list from the "proxy list 1" will be fetched.

Each hashref in the returned arrayref is in a following format
(if any field is missing on the site it will be reported as a string
C<N/A>):

    {
        'domain' => 'localfast.info',
        'rating' => '65',
        'country' => 'Germany',
        'access_time' => '1.3',
        'uptime' => '96',
        'online_since' => '16 hours',
        'last_test' => '30 minutes',
        'features_hian' => '1',
        'features_ssl' => '0',
    }

Where all the values correspond to the proxy list table columns on
the website. The C<features_hian> and C<features_ssl> keys will be
set to true values, if the proxy offers C<HiAn> or C<SSL> features
respectively.

=head2 C<error>

    my $list = $prox->get_list # just from the "proxy list 1"
        or die $prox->error;

If C<get_list()> method fails it will return
either C<undef> or an empty list, depending on the context, and the reason
for the error will be available via C<error()> method. Takes no arguments,
return a human-readable error message explaining the failure.

=head2 C<list>

    my $last_list_ref = $prox->list;

Must be called after a successfull call to C<get_list()>. Takes
no arguments, returns the same arrayref of hashrefs as
last call to C<get_list()> returned.

=head2 C<ua>

    my $old_LWP_UA_obj = $prox->ua;

    $prox->ua( LWP::UserAgent->new( timeout => 10, agent => 'foos' );

Returns a currently used L<LWP::UserAgent> object used for retrieving
data. Takes one optional argument which must be an L<LWP::UserAgent>
object, and the object you specify will be used in any subsequent calls
to C<get_list()>.

=head2 C<debug>

    my $old_debug => $prox->debug;

    $prox->debug(1);

Returns a currently set debug value, when called with an optional argument
(which can be either a true or false value) will set debug to that value.
See C<debug> argument to constructor for more information.

=head1 AUTHOR

'Zoffix, C<< <'zoffix at cpan.org'> >>
(L<http://zoffix.com>, L<http://haslayout.net>)

=head1 BUGS

Please report any bugs or feature requests to C<bug-www-proxy4freecom at rt.cpan.org>, or through
the web interface at L<http://rt.cpan.org/NoAuth/ReportBug.html?Queue=WWW-Proxy4FreeCom>.  I will be notified, and then you'll
automatically be notified of progress on your bug as I make changes.

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WWW::Proxy4FreeCom

You can also look for information at:

=over 4

=item * RT: CPAN's request tracker

L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=WWW-Proxy4FreeCom>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WWW-Proxy4FreeCom>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WWW-Proxy4FreeCom>

=item * Search CPAN

L<http://search.cpan.org/dist/WWW-Proxy4FreeCom>

=back

=head1 COPYRIGHT & LICENSE

Copyright 2008 'Zoffix, all rights reserved.

This program is free software; you can redistribute it and/or modify it
under the same terms as Perl itself.

=cut
